File: frontend/src/types/expense-item.ts
/**
 * 费用项字典类型（对齐 CONTRACTS §7）
 * 金额字段 default_qty / unit_price 为字符串化两位小数（NUMERIC，避免 JS 浮点失真）。
 */

export interface ExpenseItem {
  id: number
  org_id: number
  name: string
  /** 默认数量，字符串两位小数，如 "1.00" */
  default_qty: string
  /** 默认单价，字符串两位小数，如 "500.00" */
  unit_price: string
  /** 单位，可为 null */
  unit: string | null
  is_active: boolean
  sort_order: number
}

/** 列表查询参数（§7.1） */
export interface ExpenseItemQuery {
  org_id?: number
  is_active?: boolean
}

/** 新增请求（§7.2） */
export interface ExpenseItemCreatePayload {
  org_id: number
  name: string
  default_qty: string
  unit_price: string
  unit?: string | null
  sort_order?: number
}

/** 修改请求（§7.3，均可选） */
export interface ExpenseItemUpdatePayload {
  name?: string
  default_qty?: string
  unit_price?: string
  unit?: string | null
  is_active?: boolean
  sort_order?: number
}

File: backend/app/schemas/expense_item.py
"""
费用项字典 Schemas (CONTRACTS §7)
"""
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field


class ExpenseItemBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=128)
    default_qty: str = Field(..., pattern=r"^\d+(\.\d{1,2})?$")
    unit_price: str = Field(..., pattern=r"^\d+(\.\d{1,2})?$")
    unit: Optional[str] = Field(None, max_length=16)
    sort_order: int = Field(default=0)


class ExpenseItemCreate(ExpenseItemBase):
    org_id: int


class ExpenseItemUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=128)
    default_qty: Optional[str] = Field(None, pattern=r"^\d+(\.\d{1,2})?$")
    unit_price: Optional[str] = Field(None, pattern=r"^\d+(\.\d{1,2})?$")
    unit: Optional[str] = Field(None, max_length=16)
    is_active: Optional[bool] = None
    sort_order: Optional[int] = None


class ExpenseItemOut(BaseModel):
    id: int
    org_id: int
    name: str
    default_qty: str
    unit_price: str
    unit: Optional[str]
    is_active: bool
    sort_order: int

    model_config = {"from_attributes": True}

    @classmethod
    def from_orm_model(cls, item) -> "ExpenseItemOut":
        """
        从 ORM ExpenseItem 转换，金额字段 Decimal → 两位小数字符串。
        """
        return cls(
            id=item.id,
            org_id=item.org_id,
            name=item.name,
            default_qty=f"{item.default_qty:.2f}",
            unit_price=f"{item.unit_price:.2f}",
            unit=item.unit,
            is_active=item.is_active,
            sort_order=item.sort_order,
        )
